File: ouro_sdk_python/examples/builder_pattern.py
"""
Builder Pattern Example

Demonstrates the fluent API for creating microchains with different configurations
"""

from ouro_sdk import MicrochainBuilder, ConsensusType, AnchorFrequency


def main():
    print("🚀 Ouroboros SDK - Builder Pattern Example\n")

    # Example 1: Simple microchain
    print("📦 Creating simple microchain with builder...")
    simple = MicrochainBuilder("SimpleDApp", "ouro1owner...").node(
        "http://localhost:8001"
    ).build()

    print(f"✅ Created: {simple.id}\n")

    # Example 2: Gaming microchain with fast blocks
    print("🎮 Creating gaming microchain...")
    gaming = (
        MicrochainBuilder("GameFi", "ouro1gamedev...")
        .node("http://localhost:8001")
        .block_time(2)  # 2 second blocks for responsive gameplay
        .consensus(ConsensusType.SINGLE_VALIDATOR)  # Fast, centralized
        .anchor_frequency(AnchorFrequency.every_n_seconds(300))  # Anchor every 5 minutes
        .build()
    )

    print(f"✅ Gaming microchain: {gaming.id}")
    print("   Block time: 2 seconds")
    print("   Consensus: SingleValidator (fast)\n")

    # Example 3: DeFi microchain with high security
    print("💰 Creating DeFi microchain...")
    defi = (
        MicrochainBuilder("DeFiProtocol", "ouro1defi...")
        .node("http://localhost:8001")
        .block_time(10)  # 10 second blocks for stability
        .consensus(ConsensusType.BFT, 7)  # High security with 7 validators
        .anchor_frequency(AnchorFrequency.every_n_blocks(50))  # Frequent anchoring
        .build()
    )

    print(f"✅ DeFi microchain: {defi.id}")
    print("   Block time: 10 seconds")
    print("   Consensus: BFT with 7 validators")
    print("   Anchoring: Every 50 blocks\n")

    # Example 4: Manual anchoring microchain
    print("🔧 Creating microchain with manual anchoring...")
    manual = (
        MicrochainBuilder("ManualChain", "ouro1manual...")
        .node("http://localhost:8001")
        .anchor_frequency(AnchorFrequency.manual())
        .build()
    )

    print(f"✅ Manual microchain: {manual.id}")
    print("   Anchoring: Manual only\n")

    print("🎉 Builder pattern examples complete!")


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"❌ Error: {e}")
        exit(1)

File: ouro_sdk/examples/builder_pattern.rs
use ouro_sdk::{MicrochainBuilder, ConsensusType, AnchorFrequency};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🚀 Ouroboros SDK - Builder Pattern Example\n");

    // The builder pattern provides a fluent API for creating microchains

    // Example 1: Simple microchain
    println!("📦 Creating simple microchain with builder...");
    let simple = MicrochainBuilder::new("SimpleDApp", "ouro1owner...")
        .node("http://localhost:8001")
        .build()
        .await?;

    println!("✅ Created: {}\n", simple.id);

    // Example 2: Gaming microchain with fast blocks
    println!("🎮 Creating gaming microchain...");
    let gaming = MicrochainBuilder::new("GameFi", "ouro1gamedev...")
        .node("http://localhost:8001")
        .block_time(2)  // 2 second blocks for responsive gameplay
        .consensus(ConsensusType::SingleValidator)  // Fast, centralized
        .anchor_frequency(AnchorFrequency::EveryNSeconds(300))  // Anchor every 5 minutes
        .build()
        .await?;

    println!("✅ Gaming microchain: {}", gaming.id);
    println!("   Block time: 2 seconds");
    println!("   Consensus: SingleValidator (fast)\n");

    // Example 3: DeFi microchain with high security
    println!("💰 Creating DeFi microchain...");
    let defi = MicrochainBuilder::new("DeFiProtocol", "ouro1defi...")
        .node("http://localhost:8001")
        .block_time(10)  // 10 second blocks for stability
        .consensus(ConsensusType::Bft { validator_count: 7 })  // High security
        .anchor_frequency(AnchorFrequency::EveryNBlocks(50))  // Frequent anchoring
        .build()
        .await?;

    println!("✅ DeFi microchain: {}", defi.id);
    println!("   Block time: 10 seconds");
    println!("   Consensus: BFT with 7 validators");
    println!("   Anchoring: Every 50 blocks\n");

    // Example 4: Manual anchoring microchain
    println!("🔧 Creating microchain with manual anchoring...");
    let manual = MicrochainBuilder::new("ManualChain", "ouro1manual...")
        .node("http://localhost:8001")
        .anchor_frequency(AnchorFrequency::Manual)
        .build()
        .await?;

    println!("✅ Manual microchain: {}", manual.id);
    println!("   Anchoring: Manual only\n");

    println!("🎉 Builder pattern examples complete!");
    Ok(())
}
